l1-contracts/contracts/bridge/L1NativeTokenVault.sol

```diff
@@ -0,0 +1,290 @@
+    }
+
+    /// @dev Transfer tokens from shared bridge as part of migration process.
+    /// @param _token The address of token to be transferred (address(1) for ether and contract address for ERC20).
+    function transferFundsFromSharedBridge(address _token) external onlySelf {
+        if (_token == ETH_TOKEN_ADDRESS) {
+            uint256 balanceBefore = address(this).balance;
+            L1_SHARED_BRIDGE.transferTokenToNTV(_token);
+            uint256 balanceAfter = address(this).balance;
+            require(balanceAfter > balanceBefore, "NTV: 0 eth transferred");
+        } else {
+            uint256 balanceBefore = IERC20(_token).balanceOf(address(this));
+            uint256 sharedBridgeChainBalance = IERC20(_token).balanceOf(address(L1_SHARED_BRIDGE));
+            require(sharedBridgeChainBalance > 0, "NTV: 0 amount to transfer");
+            L1_SHARED_BRIDGE.transferTokenToNTV(_token);
+            uint256 balanceAfter = IERC20(_token).balanceOf(address(this));
+            require(balanceAfter - balanceBefore >= sharedBridgeChainBalance, "NTV: wrong amount transferred");
+        }
+    }
+
+    /// @dev Set chain token balance as part of migration process.
+    /// @param _token The address of token to be transferred (address(1) for ether and contract address for ERC20).
+    /// @param _targetChainId The chain ID of the corresponding hyperchain.
+    function transferBalancesFromSharedBridge(address _token, uint256 _targetChainId) external onlySelf {
+        uint256 sharedBridgeChainBalance = L1_SHARED_BRIDGE.chainBalance(_targetChainId, _token);
+        chainBalance[_targetChainId][_token] = chainBalance[_targetChainId][_token] + sharedBridgeChainBalance;
+        L1_SHARED_BRIDGE.transferBalanceToNTV(_targetChainId, _token);
+    }
+
+    /// @dev Transfer tokens from shared bridge as part of migration process.
+    /// @dev This method does not required access control, as the assets can only be transferred to NTV.
+    /// @dev Unlike `transferFundsFromSharedBridge` is provides a concrete limit on the gas used for the transfer and even if it will fail, it will not revert the whole transaction.
+    function safeTransferFundsFromSharedBridge(address _token, uint256 _gasPerToken) external {
+        try this.transferFundsFromSharedBridge{gas: _gasPerToken}(_token) {} catch {
+            // A reasonable amount of gas will be provided to transfer the token.
+            // If the transfer fails, we don't want to revert the whole transaction.
+        }
+    }
+
+    /// @dev Set chain token balance as part of migration process.
+    /// @dev This method does not required access control, as the balances can only be transferred to NTV.
+    /// @dev Unlike `transferBalancesFromSharedBridge` is provides a concrete limit on the gas used for the transfer and even if it will fail, it will not revert the whole transaction.
+    function safeTransferBalancesFromSharedBridge(
```

Comment:
You can remove `safeTransferBalancesFromSharedBridge` and `safeTransferFundsFromSharedBridge`. Caller can use simpler multicall (it will call it anyway to batch many calls together)